odl/contrib/theano/layer.py

```diff
@@ -17,60 +17,260 @@
+        >>> func_op = TheanoOperator(functional)
+        >>> x = theano.tensor.dvector()
+        >>> apply = func_op.make_node(x)
+        >>> apply.outputs[0].type()
+        <TensorType(float64, scalar)>
         """
         self.operator = operator
 
     def make_node(self, x):
-        """Create node in for theanos computational graph."""
+        """Create a node for the computation graph.
+
+        Parameters
+        ----------
+        x : `theano.tensor.var.TensorVariable`
+            Input to the node.
+
+        Returns
+        -------
+        node : `theano.gof.graph.Apply`
+            Node for the Theano expression graph. Its only input is ``x``,
+            and the output is of the same type.
+        """
         x = theano.tensor.as_tensor_variable(x)
 
-        # Create tensortype with correct dtype.
-        # The second argument specifies the number of dimensions of the output
+        # Create tensor type with correct dtype.
+        # The second argument specifies the number of dimensions of the output.
         # False means that we do not support broadcasting.
-        out_type = theano.tensor.TensorType(
-            self.operator.range.dtype,
-            [False] * len(self.operator.range.shape))
-        return theano.Apply(self, [x], [out_type()])
+
+        if isinstance(self.operator, Functional):
+            # Make scalar out type
+            out_type = theano.tensor.TensorVariable(
+                theano.tensor.TensorType(self.operator.domain.dtype, ()))
+        else:
+            out_type = theano.tensor.TensorVariable(
+                theano.tensor.TensorType(
+                        self.operator.range.dtype,
+                        [False] * len(self.operator.range.shape)))
+
+        return theano.Apply(self, [x], [out_type.type()])
 
     def perform(self, node, inputs, output_storage):
-        """Compute the result."""
+        """Evaluate this node's computation.
+
+        Parameters
+        ----------
+        node : `theano.gof.graph.Apply`
+            The node of this Op in the computation graph.
+        inputs : 1-element list of arrays
+            Contains an array (usually `numpy.ndarray`) of concrete values
+            supplied for the symbolic input variable ``x``.
+        output_storage : 1-element list of 1-element lists
+            The single 1-element list contained in ``output_storage``
+            by default contains only ``None``. This value must be replaced
+            by the result of the application of `odl_op`.
+
+        Examples
+        --------
+        Perform a matrix multiplication:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TheanoOperator(op)
+        >>> x = theano.tensor.dvector('x')
+        >>> op_x = matrix_op(x)
+        >>> op_func = theano.function([x], [op_x])
```

Comment:
Don't think you need parenthisis around `op_x`, removing should improve the print below

Reply:
Not quite sure what you want to say, but maybe the answer is this: `theano.function` always needs a list of inputs and a list (or dictionary) of outputs. I could add `[0]` but that wouldn't make it look more natural.

Reply:
Quite sure you can do `op_func = theano.function([x], op_x)`

as is done here: https://github.com/adler-j/odl/blob/2888f9ba927cf67e21b50c7fa75c703802b299a3/odl/contrib/theano/examples/theano_layer_matrix.py#L34